Fix part of #13764: Add validation checks for Skill misconception notes, misconception feedback, and skill content worked-examples question and answer

ashish-patwal opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13764.
2. This PR does the following: Add validation checks for `Skill misconception notes, misconception feedback, and skill content worked-examples question and answer`


## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![2021 11 13-19 31 51 screenshot](https://user-images.githubusercontent.com/63491234/141669955-2bd0b530-f45e-41b6-b3b0-22ebb73eab6e.png)
![2021 11 13-19 32 16 screenshot](https://user-images.githubusercontent.com/63491234/141669956-72867a7b-90f2-4870-b6f8-f2be8d7a3837.png)
![2021 11 13-19 32 33 screenshot](https://user-images.githubusercontent.com/63491234/141669964-a3f08a12-dd92-495a-9367-422796c6f822.png)
![2021 11 13-19 32 45 screenshot](https://user-images.githubusercontent.com/63491234/141669978-ed98cecd-2464-4b4c-bc53-af07d8cb4bfb.png)
![2021 11 13-23 39 28 screenshot](https://user-images.githubusercontent.com/63491234/141669997-8af03541-0eb4-4a2d-9949-d2f77d8df35d.png)
<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ashish-patwal committed:
initial commit

EricZLou commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.ts:
```diff
@@ -61,6 +61,13 @@ angular.module('oppia').controller('AddMisconceptionModalController', [
 
+    $scope.isEditorContentLengthExceeded = function() {
+      return (
+        $scope.misconceptionNotes.length > 1000 ||
+        $scope.misconceptionFeedback.length > 1000
```
Almost! The validation check that we want is that the length of the **HTML string** is at most 1000 characters. See the other 2 PRs in issue #13764 that deal with HTML string length limits for samples. FYI: adding a single image to the content could add hundreds of characters to the HTML string even though an image isn't really "characters".

ashish-patwal replied:
Thanks for pointing that out . I will resolve that immediately !!

ashish-patwal replied:
hey @EricZLou , wanted to cross check some things before making a PR . Since i am capturing the whole html document inside the $scope.misconceptionNotes ( \<p>Random Text\</p> ) , it will also capture other things provided by skills-editor that you mentioned like images , collapsable inputs , links , formulas . So what i wanted to ask is that how would i take measure of the input chars / content length for all these other elements . Or i should just simply exclude these items from adding to overall input chars i.e. not taking them into consideration .

EricZLou replied:
Thanks for cross checking with me! I may have missed something here, and you may have been right originally. Add an image to the misconception note and feedback, and print out `$scope.misconceptionNotes` and `$scope.misconceptionFeedback` to see if they include `<img>` tags! If they do, then you actually don't need to change your code at all. We want to capture all of the normal characters as well as images, collapsible inputs, etc (once everything is converted to an HTML string).

ashish-patwal replied:
![html_capture](https://user-images.githubusercontent.com/63491234/141732269-b8541587-97cf-42e5-bebf-4fed41101b26.png)

@EricZLou  It is just as you said , we are capturing all the html document inserted by skill editor and not some image bytes data . So i don't think there is any need to change to code . However I will update other changes that you requested .

EricZLou replied:
Awesome -- thanks @ashish-patwal for checking this!!

EricZLou commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.directive.html:
```diff
@@ -51,14 +51,17 @@ <h3>Add Misconception</h3>
         <input type="checkbox" ng-model="misconceptionMustBeAddressed">
       </div>
+      <div ng-if="isEditorContentLengthExceeded()" class="oppia-length-validation-error">
+        The content length is too long. Please use at most 1000 characters.
```
Let's make this message clearer for the creator, who very likely doesn't know that we're setting a maximum on the number of chars in the HTML string.
```suggestion
        Please shorten the content (images, tabs, and other special items take up the most storage).
```

ashish-patwal replied:
Done !! Since i am using the same line at 4 different places , should i add it to the assets/i18n/en.json .

EricZLou commented on core/templates/pages/skill-editor-page/modal-templates/add-worked-example-modal.controller.ts:
```diff
@@ -36,6 +36,13 @@ angular.module('oppia').controller('AddWorkedExampleModalController', [
     $scope.tmpWorkedExampleQuestionHtml = '';
     $scope.tmpWorkedExampleExplanationHtml = '';
 
+    $scope.isEditorContentLengthExceeded = function() {
+      return (
+        $scope.tmpWorkedExampleQuestionHtml.length > 1000 ||
+        $scope.tmpWorkedExampleExplanationHtml.length > 1000
+      );
+    };
```
I think it's a good idea to break these 2 checks into 2 different functions (and 2 different errors) as if they're completely separate! This is good because: a lot of people will see these errors as soon as we merge this PR, and it may be confusing whether the Question or the Explanation (or both) have errors.

ashish-patwal replied:
Sure , i will resolve that asap .

EricZLou reviewed (changes requested):
Thanks @ashish-patwal! Left some comments. Let me know what questions you have!

EricZLou commented:
@ashish-patwal Responded to your comment above!

ashish-patwal committed:
updated code

ashish-patwal commented:
Updated code screenshots : 
![2021 11 15-12 22 39 screenshot](https://user-images.githubusercontent.com/63491234/141739510-485730d6-c6df-4600-8758-fdc7957b3557.png)
![2021 11 15-12 22 47 screenshot](https://user-images.githubusercontent.com/63491234/141739513-89f71a6b-5394-4210-8e8b-0a64b457c045.png)
![2021 11 15-12 42 01 screenshot](https://user-images.githubusercontent.com/63491234/141739519-9d69e8c3-5b96-41a4-91c1-9d9dd9f2c9f6.png)
![2021 11 15-12 42 11 screenshot](https://user-images.githubusercontent.com/63491234/141739521-ab5f72c4-4527-4ea0-9c82-589e2e31e30b.png)
![2021 11 15-12 55 34 screenshot](https://user-images.githubusercontent.com/63491234/141739525-de3dc0f9-74ff-44d1-95a3-76374d5bc92a.png)

PTAL @EricZLou

ashish-patwal committed:
lint-checked and passed

ashish-patwal committed:
pass failed runs

ashish-patwal commented:
PTAL @EricZLou

EricZLou commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.spec.ts:
```diff
@@ -121,4 +121,26 @@ describe('Add Misconception Modal Controller', function() {
+
+  it('should check content length in misconception notes within 1000 chars',
+    function() {
+      var contentText = 'This is just a sample text.';
+
+      $scope.misconceptionNotes = contentText;
+      expect($scope.isMisconceptionNotesHtmlLengthExceeded()).toBe(false);
+
+      $scope.misconceptionNotes = contentText.repeat(50);
+      expect($scope.isMisconceptionNotesHtmlLengthExceeded()).toBe(true);
```
Might be more clear to do something like this (for all of the tests)!
```suggestion
      $scope.misconceptionNotes = 'a'.repeat(1000);
      expect($scope.isMisconceptionNotesHtmlLengthExceeded()).toBe(false);

      $scope.misconceptionNotes = 'a'.repeat(1001);
      expect($scope.isMisconceptionNotesHtmlLengthExceeded()).toBe(true);
```

ashish-patwal replied:
Sure looks good.  I will apply the changes

EricZLou replied:
Thanks for the lightning fast response @ashish-patwal!!

EricZLou reviewed (approved):
LGTM! Great work @ashish-patwal -- just a small change to the unit tests.

ashish-patwal committed:
Revert last commit

This reverts commit a1a88a0cbdaea9dd963e476d9fff7aebea55047a.

ashish-patwal committed:
unit test update

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

EricZLou commented:
@kevintab95 @aishwary023 PTAL as codeowners!

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

kevintab95 commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.ts:
```diff
@@ -61,6 +61,12 @@ angular.module('oppia').controller('AddMisconceptionModalController', [
     $scope.misconceptionNameIsDuplicate = false;
 
+    $scope.isMisconceptionNotesHtmlLengthExceeded = () =>
+      $scope.misconceptionNotes.length > 1000;
```
Here and below, please use a named constant for the number.

ashish-patwal replied:
PTAL @kevintab95

kevintab95 commented on core/templates/pages/skill-editor-page/modal-templates/add-worked-example-modal.controller.ts:
```diff
@@ -36,6 +36,12 @@ angular.module('oppia').controller('AddWorkedExampleModalController', [
     $scope.tmpWorkedExampleExplanationHtml = '';
 
+    $scope.isWorkedExampleQuestionHtmlLengthExceeded = () =>
+      $scope.tmpWorkedExampleQuestionHtml.length > 1000;
```
Here and below, please use a named constant for the number.

ashish-patwal replied:
PTAL @kevintab95

kevintab95 reviewed (changes requested):
Thanks @ashish-patwal! Just 2 nits. PTAL

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

ashish-patwal committed:
added named constants

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

aishwary023 commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.spec.ts:
```diff
@@ -121,4 +121,22 @@ describe('Add Misconception Modal Controller', function() {
     expect($uibModalInstance.close).not.toHaveBeenCalled();
   });
+
+  it('should check content length in misconception notes within 1000 chars',
```
'should check if the content length of misconception notes is within 1000 chars'. Similarly, you may write for the next test.

ashish-patwal replied:
Done @aishwary023

aishwary023 commented on core/templates/pages/skill-editor-page/modal-templates/add-worked-example-modal.controller.spec.ts:
```diff
@@ -47,4 +47,22 @@ describe('Add Worked Example Modal Controller', function() {
     });
   });
+
+  it('should check content length in example questions within 1000 chars',
```
ditto for the tests in these files.

ashish-patwal replied:
DONE @aishwary023

EricZLou commented:
@ashish-patwal when we respond to a reviewer's comments, it's helpful to **assign** them so they know to look at the PR again! You can do this by clicking on the Settings icon next to Reviewees.

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

Co-authored-by: Vasa Srijan Reddy <redvasa@amazon.com>

ashish-patwal commented:
> @ashish-patwal when we respond to a reviewer's comments, it's helpful to **assign** them so they know to look at the PR again! You can do this by clicking on the Settings icon next to Reviewees.

@EricZLou Thanks !!  I'll remember that .

ashish-patwal committed:
updated tests description

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation

ashish-patwal committed:
Merge remote-tracking branch 'refs/remotes/origin/skills-page-vaildation' into skills-page-vaildation

ashish-patwal commented:
PTAL @aishwary023

srijanreddy98 committed:
Merge branch 'develop' into skills-page-vaildation

srijanreddy98 reviewed (changes requested):
Hi @ashish-patwal. In order to remove some lines from the final commit message in this PR, I raised this "request changes" comment on this PR so that it can't be merged by anyone else. You don't have to do anything. I will be monitoring the PR and when it has all the approvals and the tests passing I will merge the PR. Please reply to the email I have sent you if you have any questions about this. Thanks -- Srijan

srijanreddy98 committed:
Merge branch 'develop' into skills-page-vaildation

kevintab95 commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.ts:
```diff
@@ -26,16 +26,20 @@ require('pages/skill-editor-page/services/skill-editor-state.service.ts');
     $scope.skill = SkillEditorStateService.getSkill();
     $scope.MAX_CHARS_IN_MISCONCEPTION_NAME =
       MAX_CHARS_IN_MISCONCEPTION_NAME;
+    $scope.MAX_CHARS_IN_SKILL_MISCONCEPTION_EDITOR =
```
Looks like we are not using `MAX_CHARS_IN_SKILL_MISCONCEPTION_EDITOR` directly in the template. In that case, let's avoid attaching it to `scope`. You can directly do `$scope.misconceptionXXX.length > MAX_CHARS_IN_SKILL_MISCONCEPTION_EDITOR` below.

ashish-patwal replied:
@kevintab95 Done !

kevintab95 commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.directive.html:
```diff
@@ -38,6 +38,9 @@ <h3>Add Misconception</h3>
         </div>
         <schema-based-editor class="protractor-test-notes-textarea" schema="MISCONCEPTION_PROPERTY_FORM_SCHEMA" local-value="misconceptionNotes">
         </schema-based-editor>
+        <div ng-if="isMisconceptionNotesHtmlLengthExceeded()" class="oppia-length-validation-error mt-2">
+           Please shorten the content (images, tabs, and other special items take up the most storage).
+        </div>
```
Hm, do we actually need to count the characters taken up by actual HTML tags? It might be hard for a user to figure out if their content is under the limit since they may not see the actual HTML tags. @EricZLou could you please check if this is expected?

ashish-patwal replied:
@kevintab95 i discussed the same thing with @EricZLou and he agrees with having html tags included in the count . You can read the top threads to get more insight into this matter .

EricZLou replied:
@kevintab95 -- good question; the purpose of having these "HTML length" validation checks it to limit the amount of data that is ultimately stored in our backend storage.

**Previously:** the user could attach 1,000 images and write 2 characters and the validation check would just see a length of 2, even though it is hundreds of thousands of characters that are stored.
**Now:** we limit the HTML length to account for this

Do you think this approach makes sense?

kevintab95 replied:
Ah, I see. I missed the previous comment / discussion. Sounds good, thanks!

nithusha21 replied:
Hmmm, echoing my comment, and I don't think I'm convinced yet... Also given this exact message, it is not trivial for the user to find how to get down to the limit. 

I don't think checking the character limit on HTML is really a good idea. The HTML is our internal representation, which can definitely change. We can change the RTE, add new fields in the HTML, or anything like that which would affect the user because of this check. Firstly could you explain why we need this check? And why it makes sense to validate the user input as HTML?

For your concern above, we can have better validations - like limit number of images in the html (or videos). But limiting on the size of the HTML doesn't sound right.

kevintab95 replied:
I agree with @nithusha21 here. Consider this scenario, the misconception is valid when a user creates it but in the future we do a migration that modifies an existing tag to have more characters than before and that results in the misconception becoming invalid (exceeds the char limit). This seems problematic -- the user won't know what caused it to become invalid.

ashish-patwal replied:
@kevintab95 @nithusha21 @EricZLou I think if we are going to make changes , we need more insight on the approach through which we are going to validate the checks i.e. what to validate checks against ( raw html data, actual content inside html tags or something else ) . Waiting for your response !!

nithusha21 replied:
Yes @ashish-patwal I definitely agree. As a reviewer I am not comfortable with the changes in this PR as of now. The impact of changing validations is usually drastic if not done correctly so we need to understand a few things. 3 things primarily, what are is the purpose of adding the validation, how can the user easily ensure the input is valid, and would the validation only affect cases when people enter a valid input and then the input becomes invalid after a change on the developer side which is hidden from the user. 

So this being a hard problem by itself, let me ask a few questions that we can try to address and then we can decide on what the right thing to do is...
-  I don't understand the purpose of this validation (as we don't seem to have similar validations in the backend right?). Are we trying to limit the user input to reduce the amount of data we store in the database? Or is there any particular reason apart from this where we want to reduce the size of inputs?
- The validation we add should strictly be on the basis of user input. A user input is either valid or invalid based on what is input. It shouldn't depend on internal representation of the data. As an example, if we would like to limit large images, validate the size of the image added (I think this validation exists). If we would like to limit number of images, limit that specifically. 
- The scheme that we come up with needs to be validated against existing data. If users have saved "invalid" data before your change, it would cause confusion as it would suddenly become invalid. We need to audit and fix these cases.
-  To reiterate, this is not an easy problem, and would not be easy to get right in 1 go. It is one of those cases where it is easy to say this approach is incorrect, but it is much harder to come up with the correct checks. 

Here is my suggestion. As this change is becoming slightly more involved lets start by making a doc (1-2 pages would suffice 😄) which addresses these points, and suggests an approach which we think would work, and you can send it to Kevin, Eric and me. We can discuss over the approach and finalize the strategy that works. Does this sound good?

kevintab95 commented on core/templates/pages/skill-editor-page/modal-templates/add-worked-example-modal.controller.ts:
```diff
@@ -22,12 +22,18 @@ require(
       $uibModalInstance: $uibModalInstance
     });
 
+    $scope.MAX_CHARS_IN_SKILL_WORKED_EXAMPLE_EDITOR =
```
Ditto: avoid attaching this to the template scope if it is not used directly in the template HTML.

ashish-patwal replied:
@kevintab95 Done !

kevintab95 reviewed (changes requested):
Thanks @ashish-patwal! Left a few comments. PTAL

ashish-patwal committed:
removed MAX_CHARS_IN_SKILL from template scope

seanlip commented:
Taking @krishnarao22 and me off, not sure why we were added by oppiabot.

EricZLou commented:
@nithusha21 @aishwary023 PTAL -- thanks!

nithusha21 commented on core/templates/pages/skill-editor-page/modal-templates/add-misconception-modal.controller.spec.ts:
```diff
@@ -121,4 +121,24 @@ describe('Add Misconception Modal Controller', function() {
+  it('should check if the content length of misconception feedback is within 1000 chars',
+    function() {
+      $scope.misconceptionFeedback = 'a'.repeat(1000);
+      expect($scope.isMisconceptionFeedbackHtmlLengthExceeded()).toBe(false);
```
I'm a bit confused... The HTML is not something the user would input directly... Putting limits on the HTML size could be confusing to the user. How do we communicate this limit to the user (on the UI)?

ashish-patwal replied:
@nithusha21 I believe you have missed some threads where we had discussion regarding the same . We came to a conclusion to limit the total html chars rather than the content inside of that html tags . It also fullproof against things like image tags where actual content of an image can break the limit instantly ( image data can contain thousands of character ) . You can go through the top threads to read about it . Thanks !!

nithusha21 replied:
I did go through the threads, and I am not convinced this is the right thing to do. Left a more detailed comment in the other thread, we can continue the discussion there.

nithusha21 reviewed (changes requested):
Do we know that this doesn't break any of the existing skills or misconceptions in production? We should definitely audit those. But larger design concerns voiced below 😄

EricZLou commented:
@ashish-patwal -- seems like this starter task is getting a bit more complicated than we anticipated! At this point, I think you've done really well on this PR and that you've explored some important parts of the codebase. I'll consider this to be your first completed starter task, and I'll take it over from here. Message me at ericzlou@gmail.com and I can help you find a second starter task. Sorry for the confusion, but thank you for working on this important issue!

@nithusha21 @kevintab95 -- thank you for bringing up your concerns! Totally agree that setting a strict HTML character limit may not be the best approach here. Will respond to your questions above soon, but I'll take over this PR from here.

nithusha21 commented:
Sounds good Eric 😄 Also @ashish-patwal I think this is a great first PR!

ashish-patwal commented:
> @ashish-patwal -- seems like this starter task is getting a bit more complicated than we anticipated! At this point, I think you've done really well on this PR and that you've explored some important parts of the codebase. I'll consider this to be your first completed starter task, and I'll take it over from here. Message me at [ericzlou@gmail.com](mailto:ericzlou@gmail.com) and I can help you find a second starter task. Sorry for the confusion, but thank you for working on this important issue!
> 
> @nithusha21 @kevintab95 -- thank you for bringing up your concerns! Totally agree that setting a strict HTML character limit may not be the best approach here. Will respond to your questions above soon, but I'll take over this PR from here.

Yeah sure !! I think i'll leave it to you guys for the better as you are more familiar with oppia . Thanks again for reviewing my work . Looking towards making some more PR' s definitely the easy ones first XD .

ashish-patwal commented:
> Sounds good Eric smile Also @ashish-patwal I think this is a great first PR!

Thanks @nithusha21 and this is actually my 4th or 5th PR for oppia . Hurray !!

ashish-patwal committed:
Merge branch 'oppia:develop' into skills-page-vaildation
